### PR TITLE
Add `direct_target` to Room struct

### LINKED ISSUE #103: Parse DirectEvent and expose info in Room struct
**Is your feature request related to a problem? Please describe.**
I'd like to handle "direct" rooms differently from channels - it would be trivial to handle these as IRC queries instead of channels and match my usage of IRC better.

**Describe the solution you'd like**
I get a "m.direct" AnyBasicEvent event on initial sync that ruma deserializes to `DirectEvent<DirectEventContent>`, defined as follow:
```rust
pub struct DirectEventContent(pub BTreeMap<UserId, Vec<RoomId>>);
```

We could add a "direct" bool (or option bool) to our `Room` struct, setting true for elements in that map when we see it. Or maybe a direct_target as `option<UserId>` since we have the info?
I'm not exactly sure what happens to the `direct` flag when you invite more people in them, let me know if you want me to check... That might lead to weird behaviours.

**Describe alternatives you've considered**
I can parse the event on my side & store the info on my per roomid hashmap; I'm not sure how useful the feature would be for other consumers of the lib.

**Additional context**
(If I get an ok-in-principle I'd be happy to try submitting a PR; not sure how long I'll laste but might as well get involved a bit more :P)

### PR DESCRIPTION
**Pull request by martinetd**
Add an optional `direct_target` field to `Room` struct to track rooms marked as "direct" : for direct rooms, it will contain the associated `UserId`.

Closes #103.

Notes:
- Unfortunately was a bit more work than I had expected because we only used to parse account events tied to rooms, not the global ones -- in particular I'm not sure how to handle the saving updated rooms for the general event, the usual "return bool" doesn't work well the helper should return a vect of updated rooms or something it's a bit annoying. Probably want to fix this.
- I exposed the RoomMember struct more visibly so users wouldn't need to explicitely declare the matrix_sdk_base crate -- not sure if we want that but it's been handy for my testing from outside
- ... speaking of which, probably want to add some test. I've seen how hand-written json samples are used at the end just not sure if we want a new test or if it could integrate in one?

### CODE REVIEW

**Comment by martinetd** [comment]
Tests failing due to type length limit; I'm not comfortable changing that in a lib (and wouldn't know where to put it anyway).

I'd say it was a preexisting problem that would deserve its own PR even if that means delaying this.

Reading up on it, it looks like "boxing" things helps split types down?
https://users.rust-lang.org/t/exponential-type-size-of-async-function/36852
But I don't think I understand most of what I'm reading about it right now, or at least how it applies to this project -- having a way to measure individual types / find where the "big types" are would help perhaps...
OTOH I like what I'm reading, I was thinking the compilation was very slow but if we improve the future chains type size it'll apparently speed up considerably, that's nice! :D

**Comment by poljar** [comment]
> Tests failing due to type length limit; I'm not comfortable changing that in a lib (and wouldn't know where to put it anyway).

I think we can just bump the limits, there are a couple of issues in rustc with this, relevant issue is over here https://github.com/rust-lang/rust/issues/54540. At the bottom mostly, there was a regression as it seems. A new Rust release is scheduled quite soon so we'll be able to remove them once this is fixed.

File: matrix_sdk_base/src/client.rs [hunk 1]
```diff
@@ -695,6 +695,25 @@ impl BaseClient {
         // }
     }
 
+    /// Handle a m.direc event, updating room states if necessary.
```
Comment (poljar) [inline matrix_sdk_base/src/client.rs:698]:
```suggestion
    /// Handle a m.direc event, updating the room states if necessary.
```

File: matrix_sdk_base/src/models/room_member.rs [hunk 2]
```diff
@@ -83,6 +83,14 @@ impl PartialEq for RoomMember {
 }
 
 impl RoomMember {
+    /// Create a new room member.
+    ///
+    /// # Arguments
+    ///
+    /// * `event` - event associated with member joining, leaving or getting
```
Comment (poljar) [inline matrix_sdk_base/src/models/room_member.rs:90]:
```suggestion
    /// * `event` - event associated with a member joining, leaving or getting
```

File: matrix_sdk_base/src/models/room_member.rs [hunk 3]
```diff
@@ -83,6 +83,14 @@ impl PartialEq for RoomMember {
...
+    /// invited to room.
```
Comment (poljar) [inline matrix_sdk_base/src/models/room_member.rs:91]:
```suggestion
    /// invited to a room.
```

File: matrix_sdk_base/src/models/room_member.rs [hunk 4]
```diff
@@ -83,6 +83,14 @@ impl PartialEq for RoomMember {
...
+    ///
+    /// * `room_id` - The unique id of the room this is member of.
```
Comment (poljar) [inline matrix_sdk_base/src/models/room_member.rs:93]:
```suggestion
    /// * `room_id` - The unique id of the room this member is part of.
```

**Review by poljar — CHANGES_REQUESTED** [review]
Generally looks good, couple of grammar issues.

As I mentioned already all of this will change in the future once we get rid of our snapshot based state store.

File: matrix_sdk_base/src/client.rs (hunk 1 above)
Comment (martinetd) [reply matrix_sdk_base/src/client.rs]: (This event is not tied to a room but multiple rooms, made rooms plural instead.)

**Comment by martinetd** [comment]
Thank you, fixed the comments in their original commits as a force push (please say for future PRs if you prefer distinct fixing commits - I try to keep commits as small as possible while ensuring each commit is independently good to allow bisection)

I've also sneaked in some serialization test fixes.


Regarding the type length - you are indeed correct this is better on nightly, I've installed the toolchain to test and it passes.
However compilation is still awfully long (to my taste) -- for example the example login program takes over 10s to build on my laptop (build all, touch login.rs, build again).. As said multiple times I'm very new to rust so I don't really know what's normal but I a have similarily-sized toy `hyper` program that only takes ~2s to build (ruma uses hyper so it won't be faster but I'm not sure x4 is warranted?)
Well, I don't know. I'll still try to play with boxes and report when I figure out how to make this work -- but that's unrelated to this PR, so later :)

Until then I'm fine with the type length annotation, I thought it was for tests but it's the examples so I've added them in a new commit prior to breakage.

File: matrix_sdk/examples/command_bot.rs [hunk 5]
```diff
@@ -1,3 +1,7 @@
+// allow biggr types until rustc fix, consider removing after new rust
+// stable release. see https://github.com/rust-lang/rust/issues/54540
+#![type_length_limit = "1075569"]
```
Comment (poljar) [inline matrix_sdk/examples/command_bot.rs:3]: This seems to be unneeded now that Rust 1.47.0 got released. :sweat_smile:

**Review by poljar — CHANGES_REQUESTED** [review]
> Thank you, fixed the comments in their original commits as a force push (please say for future PRs if you prefer distinct fixing commits - I try to keep commits as small as possible while ensuring each commit is independently good to allow bisection)

I do dislike force pushing since it's a bit easier to see the changes if no force pushing is done.

> Regarding the type length - you are indeed correct this is better on nightly, I've installed the toolchain to test and it passes.

As mentioned this seems to be fixed with Rust 1.47.0.

> However compilation is still awfully long (to my taste) -- for example the example login program takes over 10s to build on my laptop (build all, touch login.rs, build again).. As said multiple times I'm very new to rust so I don't really know what's normal but I a have similarily-sized toy `hyper` program that only takes ~2s to build (ruma uses hyper so it won't be faster but I'm not sure x4 is warranted?)

I think that's the linking step taking this long, considering that there won't be much new stuff to compile.

> Well, I don't know. I'll still try to play with boxes and report when I figure out how to make this work -- but that's unrelated to this PR, so later :)

I don't think boxing is the right approach here, using async syntax is the preferred way as the linked github issue also suggested.

File: matrix_sdk/examples/command_bot.rs (hunk 5 above)
Comment (martinetd) [reply matrix_sdk/examples/command_bot.rs:3]: meh, I just downloaded the latest stable again with rustup a couple of hours ago as I was using distro stuff before trying nightly, and it wasn't available yet!! heh.

Will the travis environment have 1.47 right away?

**Comment by martinetd** [comment]
> I do dislike force pushing since it's a bit easier to see the changes if no force pushing is done.

I've got to say that's a github problem more than a force pushing problem. gerrit, gitlab etc do a great job with force push :P 
I can leave fixup commits unsquashed and push normally then rebase/force push after you've reviewed new changes if you're ok with that?

> I think that's the linking step taking this long, considering that there won't be much new stuff to compile.

Hard to tell with cargo but that appears correct (~4s for compilation and ~9 for linking); I guess changing things there won't help much with that..

> I don't think boxing is the right approach here, using async syntax is the preferred way as the linked github issue also suggested.

It says `async fn` won't make deep types, but we obviously have the problem if we hit the 1MB limit on 'bugged' rust? I agree it was a performance regression/we shouldn't carry the length 'fix' now it's fixed but that we're hitting in in the first place makes me think we are generating deep types somewhere.
Anyway, let's drop that discussion for now, I didn't intend to take your time on that.
I've pushed a revert commit for the length type, let's see how travis fares.

File: matrix_sdk/examples/command_bot.rs (hunk 5 above)
Comment (poljar) [reply matrix_sdk/examples/command_bot.rs]: Indeed it will, it fixed the CI on the crypto-improvements branch as well.

**Comment by martinetd** [comment]
Thanks!

**Comment by poljar** [comment]
Thanks for the PR.

**Pull request merged**